accepted/2020/mono-convergence/platform-specific-httpclient.md

```diff
@@ -0,0 +1,452 @@
+
+### Handlers in Detail
+**SocketsHttpHandler Handler**
+
+The handler which is purely managed and behaves generally the same across all platforms. The same downsides as iOS apply here as well.
+
+**AndroidClientHandler Handler**
+
+The AndroidClientHandler is a bridge for Java.Net.HttpURLConnection. Not only does it provide a better performance, but it also helps drive a more consistent Android experience when it comes to networking. 
+
+## WebAssembly
+
+The web browser-based handler is the only option on this platform as no sockets are available.
+
+### Handlers in Detail
+
+**WasmHttpMessageHandler**
+
+This handler is a special bridge between C# and Javascript. As Wasm matures, the shape of the handler and what it supports will change significantly and we could also rewrite it to be WASI based.
+
+**SocketsHttpHandler Handler**
+
+It’s not supported at all due to lack of Sockets on the platform. Changing it to throw PNSE for every public member will maintain API compatibility and reduce size.
+
+
+## .NET 5 Adoption Proposals
+Platform-specific handlers are implemented on top of the enhanced platform interop layer. They are in most cases relatively complex even with such abstractions available. It’s possible to implement them fully using basic interop mechanics like pinvoke but that would be tedious and very error-prone due to duplication of all the interop subtle details Xamarin and WebAssembly SDK built over the years. We would also need to introduce Xamarin namespace and build TFM specific versions of relevant assemblies as well as convert all publicly exposes types to primitive types or types which derive only from BCL types to avoid second-level dependencies which might not be desirable.
+
+In general, keeping the platform dependant handlers away from .NET Core BCL and having them available only as part of binding APIs in optional SDKs (e.g. Xamarin.iOS) seems like the most effective integration.
+
+### Proposal #1: Use static state
+
+Extend HttpClient to be configurable via static configuration type.
+```c#
+public static class HttpClientHandlers
+{
+  public static HttpClientHandler DefaultPlatformHandler { get; internal set; }
+}
+```
+
+```c#
+public class HttpClient
+{
+  public HttpClient()
+    : this(HttpClientHandlers.DefaultPlatformHandler)
+  {
+  }
+
+  ...
+}
+```
+
+**Disadvantages**
+
+* It handles only HttpClient API and we’d need to “obsolete” manual HttpClientHandler construction.
+* Initialization of static state from Xamarin SDKs could be ugly.
+
+**Conclusion**
+
+REJECTED.  It only supports `new HttpClient()`, not `new HttpClient(new HttpClientHandler())`.
+
+### Proposal #2: Convert HttpClientHandler to Proxy pattern
+Make HttpClientHandler a partial class, make all its members virtual, and on iOS/Android have the implementation of the class wrap an HttpClientHandler instance field. The instantiation of the underlying instance will have the flexibility to be controlled by platform implementation.
```

Comment:
Why "partial"?  Making it a partial class is just a compiler trick to support various ways of structuring the code; all of the partial components would still need to be compiled in at the same time.

Reply:
Fixed